### PR TITLE
Feature/delivery-address

### PR DESCRIPTION
**Pull request by San88Francisco**

### CODE REVIEW

File: src/components/Autocomplete/AutocompleteCustom.tsx [hunk 1]
```diff
@@ -0,0 +1,74 @@
+import React, { FC, ReactNode } from 'react'
+import { Autocomplete, CircularProgress, TextField } from '@mui/material'
+import { ListboxComponent, StyledPopper } from './VariableSizeList'
+
+type PropsType = {
+  name: string
+  placeholder?: string
+  register: (name: string) => any
```
Comment (pqoqubbw) [inline src/components/Autocomplete/AutocompleteCustom.tsx]: do not use `any`

File: src/components/Autocomplete/AutocompleteCustom.tsx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  options: string[]
+  onChange: (event: any, value: string) => void
```
Comment (pqoqubbw) [inline src/components/Autocomplete/AutocompleteCustom.tsx]: same here

File: src/components/Autocomplete/AutocompleteCustom.tsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  loading: boolean
+  setValueInput: (value: string) => void
+  errors: any
+  disabled?: any
```
Comment (pqoqubbw) [inline src/components/Autocomplete/AutocompleteCustom.tsx]: same here

File: src/components/Autocomplete/AutocompleteCustom.tsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  placeholder,
+  disabled
+}) => {
+  return (
+    <Autocomplete
+      {...register(name)}
+      id="name"
+      sx={{ width: 300, mt: 2 }}
+      disableListWrap
+      onError={errors[name] ? () => {} : undefined}
```
Comment (pqoqubbw) [inline src/components/Autocomplete/AutocompleteCustom.tsx]: `() => {}` what is it?

File: src/components/Autocomplete/AutocompleteCustom.tsx [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+                {params.InputProps.endAdornment}
+              </React.Fragment>
+            )
+          }}
+        />
+      )}
+      renderOption={(props, option, state) =>
+        [props, option, state.index] as ReactNode
+      }
+      renderGroup={(params) => params as any}
```
Comment (pqoqubbw) [inline src/components/Autocomplete/AutocompleteCustom.tsx]: why do you need this?

File: src/components/Autocomplete/VariableSizeList.tsx [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  forwardRef,
+  useContext,
+  useEffect,
+  useRef
+} from 'react'
+
+const LISTBOX_PADDING = 8
+
+function renderRow(props: ListChildComponentProps) {
+  const { data, index, style } = props
```
Comment (pqoqubbw) [inline src/components/Autocomplete/VariableSizeList.tsx]:
```suggestion
function renderRow({ data, index, style }: ListChildComponentProps) {
```

File: src/components/Autocomplete/VariableSizeList.tsx [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+      {...dataSet[0]}
+      noWrap={false}
+      style={inlineStyle}
+    >
+      {dataSet[1]}
+    </Typography>
+  )
+}
+
+const OuterElementContext = createContext({})
```
Comment (pqoqubbw) [inline src/components/Autocomplete/VariableSizeList.tsx]: forgot type for context

File: src/components/Autocomplete/VariableSizeList.tsx [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+const OuterElementType = forwardRef<HTMLDivElement>((props, ref) => {
+  const outerProps = useContext(OuterElementContext)
+  return <Box ref={ref} {...props} {...outerProps} />
+})
+
+function useResetCache(data: any) {
```
Comment (pqoqubbw) [inline src/components/Autocomplete/VariableSizeList.tsx]: same issue with `any`

File: src/components/Autocomplete/VariableSizeList.tsx [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+    }
+  }, [data])
+  return ref
+}
+
+export const ListboxComponent = forwardRef<
+  HTMLDivElement,
+  HTMLAttributes<HTMLElement>
+>(function ListboxComponent(props, ref) {
+  const { children, ...other } = props
```
Comment (pqoqubbw) [inline src/components/Autocomplete/VariableSizeList.tsx]:
```suggestion
>(function ListboxComponent({ children, ...other }, ref) {
```

File: src/components/Error/Error.tsx [hunk 10]
```diff
@@ -0,0 +1,22 @@
+import { FormHelperText } from '@mui/material'
+import { FC } from 'react'
+
+export const ErrorMessage: FC<{ error: any; styles?: any }> = ({
```
Comment (pqoqubbw) [inline src/components/Error/Error.tsx]: same issue with `any`

File: src/components/Modal-custom-btn/ModalCustomFormRadius.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
+import React, { Dispatch, FC, SetStateAction } from 'react'
+import InfoConfirmationModal from './ModalCustomWindow'
+import { Typography } from '@mui/material'
+import { Button } from '../UI/Button'
+import { SubmitHandler, useForm } from 'react-hook-form'
+import { yupResolver } from '@hookform/resolvers/yup'
+import * as Yup from 'yup'
+import RadioForm from '../RadioForm/RadioForm'
+
+interface ModalCustomFormRadiusProps {
```
Comment (pqoqubbw) [inline src/components/Modal-custom-btn/ModalCustomFormRadius.tsx]: you do not need `interface` here, use `type`

File: src/components/Modal-custom-btn/ModalCustomFormRadius.tsx [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+export const AddressRetention = Yup.object().shape({
+  city: Yup.string().required('Виберіть населений пункт'),
+  warehouse: Yup.string().required('Виберіть відділення')
+})
+
+export const ModalCustomFormRadius: FC<ModalCustomFormRadiusProps> = ({
+  openModal = false,
+  setOpenModal = () => {}
```
Comment (pqoqubbw) [inline src/components/Modal-custom-btn/ModalCustomFormRadius.tsx]: do not mock functions like this

File: src/components/Modal-custom-btn/ModalCustomFormRadius.tsx [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+}) => {
+  const {
+    handleSubmit,
+    register,
+    setValue,
+    setError,
+    clearErrors,
+    formState: { errors }
+  } = useForm<any>({
```
Comment (pqoqubbw) [inline src/components/Modal-custom-btn/ModalCustomFormRadius.tsx]: same issue with `any`

File: src/components/Modal-custom-btn/ModalCustomFormRadius.tsx [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+    resolver: yupResolver(AddressRetention)
+  })
+
+  const onSubmit: SubmitHandler<any> = (data) => {
```
Comment (pqoqubbw) [inline src/components/Modal-custom-btn/ModalCustomFormRadius.tsx]: same issue with `any`

File: src/components/RadioForm/AddressNP.tsx [hunk 15]
```diff
@@ -0,0 +1,68 @@
+/* eslint-disable */
+import React, { FC, useState } from 'react'
+import { ErrorMessage } from '../Error/Error'
+import { AutocompleteCustom } from '../Autocomplete/AutocompleteCustom'
+import { useNovaPoshtaCity } from '../../hooks/useNovaPoshtaCity'
+
+type PropsType = {
+  selectedValue: any
+  errors: any
```
Comment (pqoqubbw) [inline src/components/RadioForm/AddressNP.tsx]: same issue with `any`

File: src/components/RadioForm/AddressNP.tsx [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+  setError: (name: string, error: { type: string; message: string }) => void
+  register: (name: string) => any
```
Comment (pqoqubbw) [inline src/components/RadioForm/AddressNP.tsx]: same issue with `any`

File: src/components/RadioForm/AddressNP.tsx [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    loading: loadingCity,
+    options: optionsCity,
+    cityDefault
+  } = useNovaPoshtaCity(clearErrors, setError)
+  const [settlementRef, setSettlementRef] = useState<string>(
+    'e71d9e4a-4b33-11e4-ab6d-005056801329'
+  )
+  const [street, setStreet] = useState<string>('Соборна')
+
+  const onChangeCity = (_: any, value: string) => {
```
Comment (pqoqubbw) [inline src/components/RadioForm/AddressNP.tsx]: same issue with `any`

File: src/components/RadioForm/AddressNP.tsx [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+    setValue('address_np_city', value)
+    if (cityDefault.includes(value)) {
+      clearErrors('address_np_city')
+    }
+  }
+
+  const onChangeStreet = (_: any, value: string) => {
```
Comment (pqoqubbw) [inline src/components/RadioForm/AddressNP.tsx]: same issue with `any`

File: src/components/RadioForm/RadioForm.tsx [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+import { SeparationNP } from './SeparationNP'
+import { AddressNP } from './AddressNP'
+
+type PropsType = {
+  children: ReactNode
+  register: any
+  setValue: any
+  errors: any
+  setError: any
+  clearErrors: any
```
Comment (pqoqubbw) [inline src/components/RadioForm/RadioForm.tsx]: same issue with `any`

File: src/components/RadioForm/SeparationNP.tsx [hunk 20]
```diff
@@ -0,0 +1,176 @@
+import React, { FC, useEffect, useState, useCallback, useMemo } from 'react'
+import { ErrorMessage } from '../Error/Error'
+import { AutocompleteCustom } from '../Autocomplete/AutocompleteCustom'
+import { useNovaPoshtaCity } from '../../hooks/useNovaPoshtaCity'
+import axiosInstance from '../../axios/settings'
+
+type PropsType = {
+  selectedValue: any
+  errors: any
+  register: (name: string) => any
```
Comment (pqoqubbw) [inline src/components/RadioForm/SeparationNP.tsx]: same issue with `any`

File: src/components/RadioForm/SeparationNP.tsx [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+  setValue: (name: string, value: any) => void
+  clearErrors: (name: string) => void
+}
+
+const getCityRef = (value: string, novaPoshtaCity: any[]) => {
```
Comment (pqoqubbw) [inline src/components/RadioForm/SeparationNP.tsx]: same issue with `any`

File: src/components/RadioForm/SeparationNP.tsx [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+
+  const selectedCity = novaPoshtaCity[0]?.Addresses.find(
+    (address: any) =>
+      address.MainDescription === mainDescription && address.Area === area
+  )
+
+  return selectedCity ? selectedCity.Ref : null
+}
+
+const getDefaultCityRef = (value: string, cityDefault: any[]) => {
```
Comment (pqoqubbw) [inline src/components/RadioForm/SeparationNP.tsx]: same issue with `any`

File: src/components/RadioForm/SeparationNP.tsx [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+          .replace(/\(до 10 кг\)/g, '')
+          .replace(/\(до 5 кг\)/g, '')
+          .replace(/\(до 200 кг\)/g, '')
+          .replace(/\(до 1100 кг \)/g, '')
+          .replace(/\n/g, '')
+          .replace(/№(\d+)\s*:/g, '№$1:')
+          .trim()
+          .slice(0, MAX_LENGTH) +
+        (warehouse.address_warehouse.length > MAX_LENGTH ? '...' : '')
+    )
```
Comment (pqoqubbw) [inline src/components/RadioForm/SeparationNP.tsx]: move this to `utils` as a helper functions

File: src/components/RadioForm/SeparationNP.tsx [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+  }, [warehouses])
+
+  const onChangeCity = useCallback(
+    (_event: any, value: string) => {
```
Comment (pqoqubbw) [inline src/components/RadioForm/SeparationNP.tsx]: same issue with `any`

File: src/components/RadioForm/SeparationNP.tsx [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+        setSettleRef(getCityRef(value, novaPoshtaCity))
+      } else {
+        setSettleRef(null)
+      }
+    },
+    [setValue, setValInputCity, novaPoshtaCity, cityDefault, clearErrors]
+  )
+
+  const onChangeWarehouse = useCallback(
+    (_event: any, value: string) => {
```
Comment (pqoqubbw) [inline src/components/RadioForm/SeparationNP.tsx]: same issue with `any`

File: src/hooks/useNovaPoshtaCity.ts [hunk 26]
```diff
@@ -0,0 +1,110 @@
+import { useState, useEffect, useMemo } from 'react'
+
+const cityDefault = [
```
Comment (pqoqubbw) [inline src/hooks/useNovaPoshtaCity.ts]:
```suggestion
const CITY_DEFAULT = [
```

File: src/hooks/useNovaPoshtaCity.ts [hunk 27]
```diff
@@ -0,0 +1,110 @@
...
+  { name: 'Київ', ref: 'e718a680-4b33-11e4-ab6d-005056801329' },
+  { name: 'Львів', ref: 'e71abb60-4b33-11e4-ab6d-005056801329' },
+  { name: 'Одеса', ref: 'e71c2a15-4b33-11e4-ab6d-005056801329' },
+  { name: 'Дніпро', ref: 'e717110a-4b33-11e4-ab6d-005056801329' },
+  { name: 'Харків', ref: 'e71f8842-4b33-11e4-ab6d-005056801329' },
+  { name: 'Рівне', ref: 'e71d65e1-4b33-11e4-ab6d-005056801329' }
+]
+const fetchNovaPoshtaCity = async (cityName: string) => {
+  const apiKey = 'f07607422838cfac21a0d1b8603086ca'
```
Comment (pqoqubbw) [inline src/hooks/useNovaPoshtaCity.ts]: move this to `env`

File: src/hooks/useNovaPoshtaCity.ts [hunk 28]
```diff
@@ -0,0 +1,110 @@
...
+  if (!response.ok) {
+    throw new Error('Failed to fetch data')
+  }
+
+  const data = await response.json()
+  return data.data
+}
+
+export const useNovaPoshtaCity = () => {
+  const [novaPoshtaCity, setNovaPoshtaCity] = useState<any[]>([])
```
Comment (pqoqubbw) [inline src/hooks/useNovaPoshtaCity.ts]: same issue with `any`

File: src/hooks/useNovaPoshtaCity.ts [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+        : [],
+    [novaPoshtaCity]
+  )
+  const options = useMemo(() => {
+    if (!valInputCity) return cityDefault.map((city) => city.name)
+    const filteredCities = cityDefault
+      .map((city) => city.name)
+      .filter((city) => city.toLowerCase().includes(valInputCity.toLowerCase()))
+    return filteredCities.length > 0 ? filteredCities : cityRenderNP
+  }, [valInputCity, cityRenderNP])
```
Comment (pqoqubbw) [inline src/hooks/useNovaPoshtaCity.ts]: try to simplify these functions cause now I can not understand what you do

File: src/hooks/useNovaPoshtaCity.ts [hunk 30]
```diff
@@ -0,0 +1,110 @@
...
+
+  useEffect(() => {
+    const timer = setTimeout(() => {
+      if (
+        valInputCity &&
+        valInputCity.length > 3 &&
+        !cityDefault.some((city) => city.name === valInputCity) &&
+        !/[()]/.test(valInputCity)
```
Comment (pqoqubbw) [inline src/hooks/useNovaPoshtaCity.ts]: same issue with complexity here

File: src/hooks/useNovaPoshtaCity.ts [hunk 31]
```diff
@@ -0,0 +1,110 @@
...
+      ) {
+        setNovaPoshtaCity([])
+        handleCityFetch(valInputCity)
+      }
+    }, 1000)
```
Comment (pqoqubbw) [inline src/hooks/useNovaPoshtaCity.ts]: move this number as `const`

**Review by pqoqubbw — CHANGES_REQUESTED** [review]

File: src/components/Autocomplete/AutocompleteCustom.tsx (hunk 1 above)
Comment (San88Francisco) [reply src/components/Autocomplete/AutocompleteCustom.tsx]: done

File: src/components/Autocomplete/AutocompleteCustom.tsx (hunk 2 above)
Comment (San88Francisco) [reply src/components/Autocomplete/AutocompleteCustom.tsx]: done

File: src/components/Autocomplete/AutocompleteCustom.tsx (hunk 3 above)
Comment (San88Francisco) [reply src/components/Autocomplete/AutocompleteCustom.tsx]: done

File: src/components/Autocomplete/AutocompleteCustom.tsx (hunk 4 above)
Comment (San88Francisco) [reply src/components/Autocomplete/AutocompleteCustom.tsx]: done

File: src/components/Autocomplete/AutocompleteCustom.tsx (hunk 5 above)
Comment (San88Francisco) [reply src/components/Autocomplete/AutocompleteCustom.tsx]: This renderOption allows you to customize the display of the list 
enderOption={(props, option, state) =>
        [props, option, state.index] as ReactNode
      }
      But I deleted renderGroup, that param not required

File: src/components/Autocomplete/VariableSizeList.tsx (hunk 6 above)
Comment (San88Francisco) [reply src/components/Autocomplete/VariableSizeList.tsx]: done

File: src/components/Autocomplete/VariableSizeList.tsx (hunk 7 above)
Comment (San88Francisco) [reply src/components/Autocomplete/VariableSizeList.tsx]: done

File: src/components/Autocomplete/VariableSizeList.tsx (hunk 8 above)
Comment (San88Francisco) [reply src/components/Autocomplete/VariableSizeList.tsx]: done

File: src/components/Autocomplete/VariableSizeList.tsx (hunk 9 above)
Comment (San88Francisco) [reply src/components/Autocomplete/VariableSizeList.tsx]: done

File: src/components/Modal-custom-btn/ModalCustomFormRadius.tsx (hunk 11 above)
Comment (San88Francisco) [reply src/components/Modal-custom-btn/ModalCustomFormRadius.tsx]: done

File: src/components/Modal-custom-btn/ModalCustomFormRadius.tsx (hunk 12 above)
Comment (San88Francisco) [reply src/components/Modal-custom-btn/ModalCustomFormRadius.tsx]: fine

File: src/components/RadioForm/AddressNP.tsx (hunk 16 above)
Comment (San88Francisco) [reply src/components/RadioForm/AddressNP.tsx]: done

File: src/components/RadioForm/AddressNP.tsx (hunk 17 above)
Comment (San88Francisco) [reply src/components/RadioForm/AddressNP.tsx]: done

File: src/components/RadioForm/RadioForm.tsx (hunk 19 above)
Comment (San88Francisco) [reply src/components/RadioForm/RadioForm.tsx]: done

File: src/components/RadioForm/SeparationNP.tsx (hunk 20 above)
Comment (San88Francisco) [reply src/components/RadioForm/SeparationNP.tsx]: done

File: src/components/RadioForm/SeparationNP.tsx (hunk 21 above)
Comment (San88Francisco) [reply src/components/RadioForm/SeparationNP.tsx]: done

File: src/components/RadioForm/SeparationNP.tsx (hunk 22 above)
Comment (San88Francisco) [reply src/components/RadioForm/SeparationNP.tsx]: done

File: src/components/RadioForm/SeparationNP.tsx (hunk 23 above)
Comment (San88Francisco) [reply src/components/RadioForm/SeparationNP.tsx]: done

File: src/components/RadioForm/SeparationNP.tsx (hunk 24 above)
Comment (San88Francisco) [reply src/components/RadioForm/SeparationNP.tsx]: done

File: src/components/RadioForm/SeparationNP.tsx (hunk 25 above)
Comment (San88Francisco) [reply src/components/RadioForm/SeparationNP.tsx]: done

File: src/hooks/useNovaPoshtaCity.ts (hunk 26 above)
Comment (San88Francisco) [reply src/hooks/useNovaPoshtaCity.ts]: done

File: src/hooks/useNovaPoshtaCity.ts (hunk 28 above)
Comment (San88Francisco) [reply src/hooks/useNovaPoshtaCity.ts]: done

File: src/hooks/useNovaPoshtaCity.ts (hunk 29 above)
Comment (San88Francisco) [reply src/hooks/useNovaPoshtaCity.ts]: I optimization code and deleted this fun cityRenderNP

File: src/hooks/useNovaPoshtaCity.ts (hunk 30 above)
Comment (San88Francisco) [reply src/hooks/useNovaPoshtaCity.ts]: removed

File: .env
Comment (pqoqubbw) [inline .env:1]: do not push `.env` file to git

File: src/components/Account-panel/Delivery-address/AddressDetails.tsx [hunk 32]
```diff
@@ -0,0 +1,69 @@
...
+  const cityArea = getCityArea(address.city, address.area)
+  const streetAddress = getStreetAddress(address.street, address.house_number)
+  const apartmentFloor = getApartmentFloor(
+    address.apartment_number,
+    address.floor
+  )
+
+  return `${cityArea}${cleanedAddress ? `, ${cleanedAddress}` : `, ${streetAddress}${apartmentFloor}`}`
+}
+
```
Comment (pqoqubbw) [inline src/components/Account-panel/Delivery-address/AddressDetails.tsx]: move all of this to local helper file

File: src/components/RadioForm/BranchesNP.tsx [hunk 33]
```diff
@@ -0,0 +1,164 @@
...
+    },
+    [
+      setValue,
+      setValInputCity,
+      novaPoshtaCity,
+      CITY_DEFAULT,
+      clearErrors,
+      getDefaultCityRef,
+      getCityRef,
+      setSettleRef
```
Comment (pqoqubbw) [inline src/components/RadioForm/BranchesNP.tsx]: remove all setters

File: src/components/RadioForm/PostomatsNP.tsx [hunk 34]
```diff
@@ -0,0 +1,153 @@
...
+  useEffect(() => {
+    if (isDisabled) {
+      setValue('cityPostomats', null)
+      setValue('postomats', null)
+      setValInputWarehouse('')
+      setSelectedWarehouseValue(null)
+      setNewRequest(true)
+      clearErrors('postomats')
+    }
+  }, [isDisabled, setValue, setNewRequest, clearErrors])
```
Comment (pqoqubbw) [inline src/components/RadioForm/PostomatsNP.tsx]: same here

**Review by pqoqubbw — CHANGES_REQUESTED** [review]

File: .env
Comment (pqoqubbw) [inline .env:1]: do not push this file to github, this should be local only

**Review by pqoqubbw — CHANGES_REQUESTED** [review]

File: src/hooks/useNovaPoshtaCity.ts (hunk 27 above)
Comment (San88Francisco) [reply src/hooks/useNovaPoshtaCity.ts]: done

File: .env
Comment (San88Francisco) [reply .env:1]: done

File: src/components/Account-panel/Delivery-address/AddressDetails.tsx (hunk 32 above)
Comment (San88Francisco) [reply src/components/Account-panel/Delivery-address/AddressDetails.tsx]: done

File: src/components/RadioForm/BranchesNP.tsx (hunk 33 above)
Comment (San88Francisco) [reply src/components/RadioForm/BranchesNP.tsx]: done

File: src/components/RadioForm/PostomatsNP.tsx (hunk 34 above)
Comment (San88Francisco) [reply src/components/RadioForm/PostomatsNP.tsx]: done

File: .env
Comment (San88Francisco) [reply .env:1]: done

**Review by pqoqubbw — APPROVED** [review]

**Pull request merged**